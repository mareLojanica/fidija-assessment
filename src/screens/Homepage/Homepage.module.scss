.homepage {
	width: 100%;

	&__header {
		display: flex;
		align-items: center;
		justify-content: space-between;
		width: 100%;
		margin-bottom: 1.5rem;
		border-bottom: 2px solid #ddd;
		padding-bottom: 10px;
	}

	&__title {
		font-size: 20px;
		font-weight: bold;
		color: #333;
		margin: 0;
	}

	&__grid {
		display: grid;
		grid-template-columns: repeat(auto-fill, minmax(120px, 1fr));
		gap: 20px;
		width: 100%;
		min-height: 600px;
		grid-auto-rows: 400px;

		@media (min-width: 1337px) {
			grid-template-columns: repeat(6, minmax(120px, 1fr));
		}

		@media (max-width: 1336px) {
			grid-template-columns: repeat(3, minmax(120px, 1fr));
		}

		@media (max-width: 767px) {
			grid-template-columns: repeat(2, minmax(120px, 1fr));
		}
	}

	&__loader,
	&__empty {
		display: flex;
		justify-content: center;
		align-items: center;
		margin-top: 150px;
		font-size: 16px;
		color: #777;
	}
}
