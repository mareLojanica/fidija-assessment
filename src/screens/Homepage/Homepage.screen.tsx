import React from "react";
import useHomepage from "../../hooks/useHomepageHook";
import styles from "./Homepage.module.scss";
import SuspenseLoader from "../../ui-components/SuspenseLoader";
import Header from "../../ui-components/Header";
import Main from "../../ui-components/Main";
import ShowCard from "../../ui-components/ShowCard";
import Pagination from "../../ui-components/Pagination";
import DatePicker from "../../ui-components/DatePIcker";

const Homepage: React.FC = () => {
  const {
    selectedDate,
    setSelectedDate,
    data,
    isLoading,
    pageCount,
    currentPage,
    setCurrentPage,
  } = useHomepage();

  return (
    <>
      <Header
        title="TV Blend"
        description={
          <>
            <p>TV Show and web series database.</p>
            <p>
              Create personalized schedules. Episode guide, cast, crew, and
              character information.
            </p>
          </>
        }
      />
      <Main>
        <section className={styles["homepage"]}>
          <div className={styles["homepage__header"]}>
            <h2 className={styles["homepage__title"]}>Last Added Shows</h2>
            <DatePicker
              selectedDate={selectedDate}
              onDateChange={setSelectedDate}
            />
          </div>

          {isLoading ? (
            <div className={styles["homepage__loader"]}>
              <SuspenseLoader />
            </div>
          ) : data.length ? (
            <>
              <div className={styles["homepage__grid"]}>
                {data.map((show) => (
                  <ShowCard
                    key={`${show.name}_${show.id}`}
                    show={{
                      id: show.show.id,
                      name: show.name,
                      image: show.image?.medium || "",
                      rating: show.rating.average ?? 0,
                    }}
                  />
                ))}
              </div>
              <Pagination
                pageCount={pageCount}
                currentPage={currentPage}
                onPageChange={setCurrentPage}
              />
            </>
          ) : (
            <div className={styles["homepage__empty"]}>
              <p>No shows available</p>
            </div>
          )}
        </section>
      </Main>
    </>
  );
};

export default Homepage;
