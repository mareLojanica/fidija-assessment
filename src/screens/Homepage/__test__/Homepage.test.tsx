import React from "react";
import { render, screen } from "@testing-library/react";
import Homepage from "../Homepage.screen";
import "@testing-library/jest-dom";
import useHomepage from "../../../hooks/useHomepageHook";

jest.mock("../../../hooks/useHomepageHook", () => ({
  __esModule: true,
  default: jest.fn(),
}));

jest.mock("../../../ui-components/SuspenseLoader/", () =>
  jest.fn(() => <div data-testid="suspense-loader">Loading...</div>)
);

jest.mock("../../../ui-components/Header/", () =>
  jest.fn(({ title }) => <header data-testid="header">{title}</header>)
);

jest.mock("../../../ui-components/Main", () =>
  jest.fn(({ children }) => <main data-testid="main">{children}</main>)
);

jest.mock("../../../ui-components/DatePIcker", () =>
  jest.fn(({ selectedDate, onDateChange }) => (
    <div data-testid="date-picker">
      <p>Selected Date: {selectedDate}</p>
      <button onClick={() => onDateChange("2024-01-01")}>Change Date</button>
    </div>
  ))
);

jest.mock("../../../ui-components/ShowCard", () =>
  jest.fn(({ show }) => (
    <div data-testid="show-card">
      <p>{show.name}</p>
    </div>
  ))
);

jest.mock("../../../ui-components/Pagination", () =>
  jest.fn(({ currentPage, pageCount, onPageChange }) => (
    <div data-testid="pagination">
      <p>
        Page {currentPage} of {pageCount}
      </p>
      <button
        disabled={currentPage === 1}
        onClick={() => onPageChange(currentPage - 1)}
      >
        Prev
      </button>
      <button
        disabled={currentPage === pageCount}
        onClick={() => onPageChange(currentPage + 1)}
      >
        Next
      </button>
    </div>
  ))
);

describe("Homepage Component", () => {
  beforeEach(() => {
    (useHomepage as jest.Mock).mockReturnValue({
      selectedDate: "2024-01-01",
      setSelectedDate: jest.fn(),
      data: [],
      isLoading: false,
      pageCount: 1,
      currentPage: 1,
      setCurrentPage: jest.fn(),
    });
  });

  test("renders header and main layout", () => {
    render(<Homepage />);

    expect(screen.getByTestId("header")).toBeInTheDocument();
    expect(screen.getByTestId("main")).toBeInTheDocument();
  });

  test("renders date picker", () => {
    render(<Homepage />);

    expect(screen.getByTestId("date-picker")).toBeInTheDocument();
    expect(screen.getByText("Selected Date: 2024-01-01")).toBeInTheDocument();
  });

  test("renders loading state", () => {
    (useHomepage as jest.Mock).mockReturnValue({
      selectedDate: "2024-01-01",
      setSelectedDate: jest.fn(),
      data: [],
      isLoading: true,
      pageCount: 1,
      currentPage: 1,
      setCurrentPage: jest.fn(),
    });

    render(<Homepage />);
    expect(screen.getByTestId("suspense-loader")).toBeInTheDocument();
  });

  test("renders show cards when data is available", () => {
    (useHomepage as jest.Mock).mockReturnValue({
      selectedDate: "2024-01-01",
      setSelectedDate: jest.fn(),
      data: [
        {
          id: 1,
          name: "Breaking Bad",
          show: { id: 1, image: { medium: "https://example.com/bb.jpg" } },
          rating: { average: 9.5 },
        },
        {
          id: 2,
          name: "Stranger Things",
          show: { id: 2, image: { medium: "https://example.com/st.jpg" } },
          rating: { average: 8.7 },
        },
      ],
      isLoading: false,
      pageCount: 2,
      currentPage: 1,
      setCurrentPage: jest.fn(),
    });

    render(<Homepage />);
    expect(screen.getAllByTestId("show-card")).toHaveLength(2);
    expect(screen.getByText("Breaking Bad")).toBeInTheDocument();
    expect(screen.getByText("Stranger Things")).toBeInTheDocument();
  });

  test("renders pagination when multiple pages exist", () => {
    (useHomepage as jest.Mock).mockReturnValue({
      selectedDate: "2024-01-01",
      setSelectedDate: jest.fn(),
      data: [
        {
          id: 1,
          name: "Breaking Bad",
          show: { id: 1, image: { medium: "https://example.com/bb.jpg" } },
          rating: { average: 9.5 },
        },
      ],
      isLoading: false,
      pageCount: 2,
      currentPage: 1,
      setCurrentPage: jest.fn(),
    });

    render(<Homepage />);
    expect(screen.getByTestId("pagination")).toBeInTheDocument();
    expect(screen.getByText("Page 1 of 2")).toBeInTheDocument();
  });

  test("renders 'No shows available' message when data is empty", () => {
    (useHomepage as jest.Mock).mockReturnValue({
      selectedDate: "2024-01-01",
      setSelectedDate: jest.fn(),
      data: [],
      isLoading: false,
      pageCount: 1,
      currentPage: 1,
      setCurrentPage: jest.fn(),
    });

    render(<Homepage />);
    expect(screen.getByText("No shows available")).toBeInTheDocument();
  });

  test("matches snapshot", () => {
    const { asFragment } = render(<Homepage />);
    expect(asFragment()).toMatchSnapshot();
  });
});
