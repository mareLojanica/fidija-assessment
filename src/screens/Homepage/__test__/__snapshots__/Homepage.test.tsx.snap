// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Homepage Component matches snapshot 1`] = `
<DocumentFragment>
  <header
    data-testid="header"
  >
    TV Blend
  </header>
  <main
    data-testid="main"
  >
    <section
      class="homepage"
    >
      <div
        class="homepage__header"
      >
        <h2
          class="homepage__title"
        >
          Last Added Shows
        </h2>
        <div
          data-testid="date-picker"
        >
          <p>
            Selected Date: 2024-01-01
          </p>
          <button>
            Change Date
          </button>
        </div>
      </div>
      <div
        class="homepage__empty"
      >
        <p>
          No shows available
        </p>
      </div>
    </section>
  </main>
</DocumentFragment>
`;
